src/http.nim

```diff
@@ -0,0 +1,250 @@
+import chronos, chronos/timer, httputils, chronicles, uri, tables, strutils
+
+const
+  MaxHttpHeadersSize = 8192       # maximum size of HTTP headers in octets
+  MaxHttpRequestSize = 128 * 1024 # maximum size of HTTP body in octets
+  HttpHeadersTimeout = timer.seconds(120) # timeout for receiving headers (120 sec)
+  CRLF* = "\r\n"
+  HeaderSep = @[byte('\c'), byte('\L'), byte('\c'), byte('\L')]
+
+type
+  HttpClient* = ref object
+    connected: bool
+    currentURL: Uri      ## Where we are currently connected.
+    headers: HttpHeaders ## Headers to send in requests.
+    transp*: StreamTransport
+    buf: seq[byte]
+
+  HttpHeaders* = object
+    table*: TableRef[string, seq[string]]
+
+  ReqStatus = enum
+    Success, Error, ErrorFailure
+
+  AsyncCallback = proc (transp: StreamTransport,
+      header: HttpRequestHeader): Future[void] {.closure, gcsafe.}
+  HttpServer* = ref object of StreamServer
+    callback: AsyncCallback
+
+proc recvData(transp: StreamTransport): Future[seq[byte]] {.async.} =
+  var buffer = newSeq[byte](MaxHttpHeadersSize)
+  var error = false
+  try:
+    let hlenfut = transp.readUntil(addr buffer[0], MaxHttpHeadersSize,
+        sep = HeaderSep)
+    let ores = await withTimeout(hlenfut, HttpHeadersTimeout)
+    if not ores:
+      # Timeout
+      debug "Timeout expired while receiving headers",
+             address = transp.remoteAddress()
+      error = true
+    else:
+      let hlen = hlenfut.read()
+      buffer.setLen(hlen)
+  except TransportLimitError:
+    # size of headers exceeds `MaxHttpHeadersSize`
+    debug "Maximum size of headers limit reached",
+          address = transp.remoteAddress()
+    error = true
+  except TransportIncompleteError:
+    # remote peer disconnected
+    debug "Remote peer disconnected", address = transp.remoteAddress()
+    error = true
+  except TransportOsError as exc:
+    debug "Problems with networking", address = transp.remoteAddress(),
+          error = exc.msg
+    error = true
+
+  if error:
+    buffer.setLen(0)
+  return buffer
+
+proc newConnection(client: HttpClient, url: Uri) {.async.} =
```

Comment:
By convention, `newXXX` is used in place of constructors (tho, this is also discouraged by the most recent style guide). For the sake of avoiding confusion, I would just call this method `connect()`.